_posts/2020-01-22-speed-up-rspec-boot-time.md

```diff
@@ -0,0 +1,146 @@
+
+## How to Profile Start Up Speed
+
+We first need to profile and see what's happening before we can fix it, [this](https://mildlyinternet.com/code/profiling-rails-boot-time.html) handy snippet lets you plug in directly to the require that's happens whenever any rails app is booted up including when you run RSpec.
+
+Another option is the [Bumbler gem](https://github.com/nevir/Bumbler), if the snippet doesn't work for you.
+
+The first step was to build to build an empty spec file, so that we can check purely the RSpec loading and run time.
+
+```ruby
+require 'rails_helper'
+
+class Empty
+  def self.test
+    true
+  end
+end
+
+describe Empty do
+  it { expect(Empty.test).to be_truthy }
+end
+```
+
+Now if we run RSpec, we see this large block of everything that's being loaded, and we see how it's adding up into the 10+ seconds that it takes to run a single spec.
+
+![alt text]({{ site.url }}/assets/images/2020-01-22-speed-up-rspec-boot-time/all_gems_loading.png "All gemfiles that are loaded on every spec run.")
+
+Here's what our current situation looks like, and this has been measured by taking the average of 5 runs of the empty spec file from above.
+
+`bundle exec rspec specs/benchmarks/empty_spec.rb`
+`Finished in 0.634441 seconds (files took 13.00 seconds to load)`
+
+There must be a way to either make this better, or cache it.
+
+The first option that popped to mind or going through the Gemfile, and adding `require: false` to the gems that didn't need to be loaded every time, and then specifically require them in the app. That could've worked, but honestly it's a lot of work, and possibly wouldn't bring the speed that much down.
+
+The more efficient option felt like using [Spring](https://github.com/rails/spring) to cache the app's dependencies and preload it, but if that doesn't work then it would be possible to start removing the longest to load ones.
+
+## Adding Spring
+
+This part is pretty straightforward, I based a lot of what I did off of this [other blog post](https://schwad.github.io/ruby/rails/testing/2017/08/14/50-times-faster-rspec-loading.html) and the [spring documentation](https://github.com/rails/spring).
+
+
+Step 1 is to add these two gems to the gemfile:
+
+```ruby
+group :test, :development do
+  # ...
+  gem "spring"
+  gem 'spring-commands-rspec'
+end
+```
+
+And to `bundle install` and run `bundle exec spring binstub rspec` to create `bin/rspec`.
+
+Now, lets try this again and look at the load time, using the spring command:
+
+`spring rspec spec/benchmarks/empty_spec.rb`
+`Finished in 1 second (files took 0.70336 seconds to load)`
+
+Great, it's loading so quickly! Wow! Except this would be a pretty boring blog post if the take away was "Just Use Spring". 
+Look at the run time for this same spec, it's now taking almost an extra 40% of time!
+
+This is where it gets really interesting. We need to profile our specs themselves and see where the issues lay.
```

Comment:
```suggestion
This is where it gets really interesting. We need to profile our specs themselves and see where the issues lie.
```
https://ell.stackexchange.com/questions/35162/the-reason-lays-or-lies-in-the-facts